File: 5. Imported Model/MetalTryOut-Objc/Main/MetalView.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  View for Metal Sample Code. Manages screen drawable framebuffers and expects a delegate to repond to render commands to perform drawing.
  
*/

#import "MetalView.h"

@implementation MetalView
{
@private
    __weak CAMetalLayer *_metalLayer;
    
    BOOL _layerSizeDidUpdate;
    
    id <MTLTexture>  _depthTex;
    id <MTLTexture>  _stencilTex;
    id <MTLTexture>  _msaaTex;
}
@synthesize currentDrawable      = _currentDrawable;
@synthesize renderPassDescriptor = _renderPassDescriptor;

+ (Class) layerClass
{
    return [CAMetalLayer class];
}

- (void) initCommon
{
    self.opaque          = YES;
    self.backgroundColor = nil;
    
    // setting this to yes will allow Main thread to display framebuffer when
    // view:setNeedDisplay: is called by main thread
    _metalLayer = (CAMetalLayer *)self.layer;
    
    self.contentScaleFactor = [UIScreen mainScreen].scale;
    
    _metalLayer.presentsWithTransaction = NO;
    _metalLayer.drawsAsynchronously     = YES;
    
    _device = MTLCreateSystemDefaultDevice();
    
    _metalLayer.device          = _device;
    _metalLayer.pixelFormat     = MTLPixelFormatBGRA8Unorm;
    _metalLayer.framebufferOnly = YES;
    
    _drawableSize = self.bounds.size;
}

- (id) initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
	if(self)
    {
        [self initCommon];
    }
    
    return self;
}

- (instancetype) initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    
    if(self)
    {
        [self initCommon];
    }
    return self;
}

- (void) releaseTextures
{
    _depthTex   = nil;
    _stencilTex = nil;
    _msaaTex    = nil;
}

- (void) setupRenderPassDescriptorForTexture:(id <MTLTexture>) texture
{
    // create lazily
    if (_renderPassDescriptor == nil)
        _renderPassDescriptor = [MTLRenderPassDescriptor renderPassDescriptor];
    
    // create a color attachment every frame since we have to recreate the texture every frame
    MTLRenderPassAttachmentDescriptor *colorAttachment = [MTLRenderPassAttachmentDescriptor new];
    colorAttachment.texture = texture;
    
    // make sure to clear every frame for best performance
    [colorAttachment setLoadAction:MTLLoadActionClear];
    [colorAttachment setClearValue:MTLClearValueMakeColor(0.65f, 0.65f, 0.65f, 1.0f)];
    
    // if sample count is greater than 1, render into using MSAA, then resolve into our color texture
    if(_sampleCount > 1)
    {
        BOOL doUpdate =     ( _msaaTex.width       != texture.width  )
                        ||  ( _msaaTex.height      != texture.height )
                        ||  ( _msaaTex.sampleCount != _sampleCount   );
        
        if(!_msaaTex || (_msaaTex && doUpdate))
        {
            MTLTextureDescriptor* desc = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat: MTLPixelFormatBGRA8Unorm
                                                                                            width: texture.width
                                                                                           height: texture.height
                                                                                        mipmapped: NO];
            desc.textureType = MTLTextureType2DMultisample;
            
            // sample count was specified to the view by the renderer.
            // this must match the sample count given to any pipeline state using this render pass descriptor
            desc.sampleCount = _sampleCount;
            
            _msaaTex = [_device newTextureWithDescriptor: desc];
        }
        
        // When multisampling, perform rendering to _msaaTex, then resolve
        // to 'texture' at the end of the scene
        [colorAttachment setTexture: _msaaTex];
        [colorAttachment setResolveTexture: texture];
        
        // set store action to resolve in this case
        [colorAttachment setStoreAction: MTLStoreActionMultisampleResolve];
    }
    else
    {
        // store only attachments that will be presented to the screen, as in this case
        [colorAttachment setStoreAction: MTLStoreActionStore];
    } // color0
    
    // set the newly created attachment on the descriptor as color attachment at index 0
    [_renderPassDescriptor.colorAttachments setObject:colorAttachment atIndexedSubscript:0];
    
    
    // Now create the depth and stencil attachments
    
    if(_depthPixelFormat != MTLPixelFormatInvalid)
    {
        BOOL doUpdate =     ( _depthTex.width       != texture.width  )
                        ||  ( _depthTex.height      != texture.height )
                        ||  ( _depthTex.sampleCount != _sampleCount   );
        
        if(!_depthTex || doUpdate)
        {
            //  If we need a depth texture and don't have one, or if the depth texture we have is the wrong size
            //  Then allocate one of the proper size
            MTLTextureDescriptor* desc = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat: _depthPixelFormat
                                                                                            width: texture.width
                                                                                           height: texture.height
                                                                                        mipmapped: NO];
            
            desc.textureType = (_sampleCount > 1) ? MTLTextureType2DMultisample : MTLTextureType2D;
            desc.sampleCount = _sampleCount;
            
            _depthTex = [_device newTextureWithDescriptor: desc];
        
            MTLRenderPassAttachmentDescriptor *depthAttachment = [MTLRenderPassAttachmentDescriptor new];
            depthAttachment.texture = _depthTex;
            [depthAttachment setLoadAction:MTLLoadActionClear];
            [depthAttachment setClearValue:MTLClearValueMakeDepth(1.0)];
            [depthAttachment setStoreAction: MTLStoreActionDontCare];
            
            _renderPassDescriptor.depthAttachment = depthAttachment;
        }
    } // depth
    
    if(_stencilPixelFormat != MTLPixelFormatInvalid)
    {
        BOOL doUpdate  =    ( _stencilTex.width       != texture.width  )
                        ||  ( _stencilTex.height      != texture.height )
                        ||  ( _stencilTex.sampleCount != _sampleCount   );
        
        if(!_stencilTex || doUpdate)
        {
            //  If we need a stencil texture and don't have one, or if the depth texture we have is the wrong size
            //  Then allocate one of the proper size
            
            MTLTextureDescriptor* desc = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat: _stencilPixelFormat
                                                                                            width: texture.width
                                                                                           height: texture.height
                                                                                        mipmapped: NO];
            
            desc.textureType = (_sampleCount > 1) ? MTLTextureType2DMultisample : MTLTextureType2D;
            desc.sampleCount = _sampleCount;
            
            _stencilTex = [_device newTextureWithDescriptor: desc];
        
            MTLRenderPassAttachmentDescriptor* stencilAttachment = [MTLRenderPassAttachmentDescriptor new];
            stencilAttachment.texture = _stencilTex;
            [stencilAttachment setLoadAction:MTLLoadActionClear];
            [stencilAttachment setClearValue:MTLClearValueMakeStencil(0)];
            [stencilAttachment setStoreAction: MTLStoreActionDontCare];
            
            _renderPassDescriptor.stencilAttachment = stencilAttachment;

        }
    } //stencil
}

- (MTLRenderPassDescriptor *) renderPassDescriptor
{
    id <CAMetalDrawable> drawable = self.currentDrawable;
    if(!drawable)
    {
        NSLog(@">> ERROR: Failed to get a drawable!");
        _renderPassDescriptor = nil;
    }
    else
    {
        [self setupRenderPassDescriptorForTexture: drawable.texture];
    }
    
    return _renderPassDescriptor;
}


- (id <CAMetalDrawable>) currentDrawable
{
    while (_currentDrawable == nil)
    {
        _currentDrawable = [_metalLayer newDrawable];
        
        if(!_currentDrawable)
        {
            NSLog(@"CurrentDrawable is nil");
        }
    }
    
    return _currentDrawable;
}

- (void) display
{
    // Create autorelease pool per frame to avoid possible deadlock situations
    // because there are 3 CAMetalDrawables sitting in an autorelease pool.
    
    @autoreleasepool
    {
        if(_layerSizeDidUpdate)
        {
            _drawableSize = self.bounds.size;
            
            _drawableSize.width  *= self.contentScaleFactor;
            _drawableSize.height *= self.contentScaleFactor;
            
            _metalLayer.drawableSize = _drawableSize;
            
            [_delegate reshape:self];
            
            _layerSizeDidUpdate = NO;
        }
        
        // draw
        [self.delegate render:self];
        
        _currentDrawable    = nil;
    }
}

- (void) setContentScaleFactor:(CGFloat)contentScaleFactor
{
    [super setContentScaleFactor:contentScaleFactor];
    
    _layerSizeDidUpdate = YES;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    _layerSizeDidUpdate = YES;
}

@end

File: 5. Imported Model/MetalTryOut-Objc/Main/MetalView.swift
//
//  MetalView.swift
//  Triangle
//
//  Created by Andrew K. on 6/24/14.
//  Copyright (c) 2014 Andrew Kharchyshyn. All rights reserved.
//

import UIKit
import QuartzCore
import Metal

@objc protocol MetalViewProtocol
{
    func render(metalView : MetalView)
    func reshape(metalView : MetalView)
}

@objc class MetalView: UIView
{
    //Public API
    
    var metalViewDelegate: MetalViewProtocol?
    var frameBuffer: AnyObject!
    
    func setFPSLabelHidden(hidden: Bool){
        if let label = fpsLabel{
            label.hidden = hidden
        }
    }
    
    init(frame: CGRect) {
        super.init(frame: frame)
        // Initialization code
        setup()
    }
    init(coder aDecoder: NSCoder!){
        super.init(coder: aDecoder)
        setup()
    }
    
    func display()
    {
        var frameBuf = frameBuffer as FrameBuffer
        
        frameBuf.drawableSize = self.bounds.size
        
        frameBuf.drawableSize.width  *= self.contentScaleFactor;
        frameBuf.drawableSize.height *= self.contentScaleFactor;
        
        var size = self.bounds.size
        
        size.width  *= self.contentScaleFactor
        size.height *= self.contentScaleFactor
        
        frameBuf.displayWithDrawableSize(size)
    }
    
    
    //Private
    var lastFrameTimestamp: CFTimeInterval?
    var _metalLayer: CAMetalLayer!
    var fpsLabel: UILabel?
    
    
    override class func layerClass() -> AnyClass{
        return CAMetalLayer.self
    }
    override func layoutSubviews(){
        var rightConstraint = NSLayoutConstraint(item: fpsLabel, attribute: .Right, relatedBy: .Equal, toItem: self, attribute: .Right, multiplier: 1.0, constant: 0.0)
        var botConstraint = NSLayoutConstraint(item: fpsLabel, attribute: .Bottom, relatedBy: .Equal, toItem: self, attribute: .Bottom, multiplier: 1.0, constant: 0.0)
        var heightConstraint = NSLayoutConstraint(item: fpsLabel, attribute: .Height, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 25.0)
        var widthConstraint = NSLayoutConstraint(item: fpsLabel, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 60.0)
        
        self.addConstraints([rightConstraint,botConstraint,widthConstraint,heightConstraint])
        
        (frameBuffer as FrameBuffer).layerSizeDidUpdate = true
        
        super.layoutSubviews()
    }
    
    func setup(){
    
        fpsLabel = UILabel(frame: CGRectZero)
        fpsLabel!.setTranslatesAutoresizingMaskIntoConstraints(false)
        fpsLabel!.backgroundColor = UIColor(red: 1.0, green: 1.0, blue: 1.0, alpha: 0.6)
        self.addSubview(fpsLabel)

        
        self.opaque          = true
        self.backgroundColor = nil
        
        // setting this to yes will allow Main thread to display framebuffer when
        // view:setNeedDisplay: is called by main thread
        _metalLayer = self.layer as CAMetalLayer
        
        self.contentScaleFactor = UIScreen.mainScreen().scale
        
        _metalLayer.presentsWithTransaction = false
        _metalLayer.drawsAsynchronously     = true
        
        var device = MTLCreateSystemDefaultDevice()
        
        _metalLayer.device          = device
        _metalLayer.pixelFormat     = MTLPixelFormat.BGRA8Unorm
        _metalLayer.framebufferOnly = true
        
        frameBuffer = FrameBuffer(metalView: self)
    }
}
